google-cloud-spanner/src/test/java/com/google/cloud/spanner/ValueBinderTest.java

```diff
@@ -110,15 +110,23 @@ public void reflection() throws InvocationTargetException, IllegalAccessExceptio
         // Test unary null.
         if (!binderMethod.getParameterTypes()[0].isPrimitive()) {
           Value expected = (Value) method.invoke(Value.class, (Object) null);
+          // Special case for json as it maps to ValueBinder.to(String)
+          if (method.getName().toLowerCase().equals("json")) {
```

Comment:
Did you see https://github.com/googleapis/java-spanner/pull/872#issuecomment-817587906? :slightly_smiling_face: 

Since the type of the input into the DB will be of type Json rather than String, the backend wouldn't be affected. So it's just down to the customers' usage. The result type for JSON will be String so there won't be a difference between string and json values being returned, do you think that would cause issues for the customers?

Reply:
Yes, but maybe I didn't understand it correctly (or maybe I'm confused by the `Special case for json as it maps to ValueBinder.to(String)` comment in this method). According to https://github.com/googleapis/java-spanner/pull/872#issuecomment-817587906 a user should use `to(Value.json("json"))` for JSON parameters. According to the comment in this test method, it maps to `ValueBinder.to(String)`, which would mean that the parameter would have type `String`. My worry was that that would mean that the backend would also not check that the parameter value is valid JSON.

Reply:
Ah I see what you mean now. The special case of `ValueBinder.to(String)` for JSON only applies to this test, because this test uses reflection to get and invoke the methods. It's comparing the output of `method` (`public static Value json(java.lang.String)`) with the output of `binderMethod` (`public Object ValueBinder.to(String)`, instead of `ValueBinder.to(Value.json(String))`) since the test uses the parameter type of the `method` to determine the `binderMethod`. This wouldn't happen when used by actual users though, as the users should do `ValueBinder.to(Value.json(String))`. 

I changed the test just now, to actually use the `to(Value)` binderMethod for the JSON case. Hopefully this makes what I'm trying to clearer. And it actually tests the JSON value binding now rather than testing String input again.

Reply:
Thanks for the explanation, now I understand what is going on.